src/lib.rs

```diff
@@ -207,12 +208,64 @@ where
                             s.annotations.clone(),
                         ) {
                             Ok(response) => {
-                                if !container_image.contains(response.digest.as_str()) {
-                                    let image_with_digest =
-                                        [container_image.as_str(), response.digest.as_str()]
-                                            .join("@");
-                                    container_with_images_digests[i]
-                                        .set_image(Some(image_with_digest));
+                                if add_digest_if_not_present(
+                                    container_image.as_str(),
+                                    response.digest.as_str(),
+                                    &mut container_with_images_digests[i],
+                                ) {
+                                    is_modified_with_digest = true;
+                                }
+                            }
+                            Err(e) => {
+                                policy_verification_errors.push(format!(
+                                    "verification of image {} failed: {}",
+                                    container_image, e
+                                ));
+                            }
+                        }
+                    }
+                }
+                Signature::KeylessPrefix(s) => {
+                    // verify if the name matches the image name provided
+                    if WildMatch::new(s.image.as_str()).matches(container_image.as_str()) {
+                        match verify_keyless_prefix_match(
+                            container_image.as_str(),
+                            s.keyless_prefix.clone(),
+                            s.annotations.clone(),
+                        ) {
+                            Ok(response) => {
+                                if add_digest_if_not_present(
+                                    container_image.as_str(),
+                                    response.digest.as_str(),
+                                    &mut container_with_images_digests[i],
+                                ) {
+                                    is_modified_with_digest = true;
+                                }
+                            }
+                            Err(e) => {
+                                policy_verification_errors.push(format!(
+                                    "verification of image {} failed: {}",
+                                    container_image, e
+                                ));
+                            }
+                        }
+                    }
```

Comment:
nitpicking, there's a lot of repetition going on inside of the outer `match` statement (the one looking at `Signaure` enums).

Maybe we could make the whole outer match a bit smaller by DRYing the inner matches, more specifically the code I'm highlighting with this comment.

Maybe we could have something like:

```rust
/// I've just invented some types, but I guess you understand what I meant
fn handle_verification_response(Result<Response>, &mut Vec<errors>) -> bool {}

Signature::KeylessPrefix(s) => {
                    // verify if the name matches the image name provided
                    if WildMatch::new(s.image.as_str()).matches(container_image.as_str()) {
                      is_modified_with_digest = handle_verification_response(verify_keyless_prefix_match(
                            container_image.as_str(),
                            s.keyless_prefix.clone(),
                            s.annotations.clone(),
                        ), &mut policy_verification_errors)
                    }
```
Something along these lines